db_test.go

```diff
@@ -800,46 +805,69 @@ func TestTombstoneCleanFail(t *testing.T) {
 		}
 		blockDir := filepath.Join(db.Dir(), uid.String())
 		block := createEmptyBlock(t, blockDir, meta)
+
+		// Add some some fake tombstones to trigger the compaction.
 		tomb := memTombstones{}
 		tomb[0] = Intervals{{0, 1}}
 		block.tombstones = tomb
 
 		db.blocks = append(db.blocks, block)
 		expectedBlockDirs = append(expectedBlockDirs, blockDir)
 	}
-	db.compactor = &mockCompactorFailing{t: t}
 
-	// This should fail as we are using the failing compactor!
+	// Initialize the mockCompactorFailing with a room for a single compaction iteration.
+	// mockCompactorFailing will fail on the second iteration so we can check if the cleanup works as expected.
+	db.compactor = &mockCompactorFailing{
+		t:      t,
+		blocks: db.blocks,
+		max:    totalBlocks + 1,
+	}
+
+	// The compactor should trigger a failure here.
 	testutil.NotOk(t, db.CleanTombstones())
 
+	// Now check that the CleanTombstones didn't leave any blocks behind after a failure.
 	actualBlockDirs, err := blockDirs(db.dir)
 	testutil.Ok(t, err)
 	testutil.Equals(t, expectedBlockDirs, actualBlockDirs)
 }
 
+// mockCompactorFailing creates a new empty block on every write and fails when reached the max allowed total.
 type mockCompactorFailing struct {
 	t      *testing.T
 	blocks []*Block
+	max    int
 }
 
 func (*mockCompactorFailing) Plan(dir string) ([]string, error) {
-
 	return nil, nil
 }
 func (c *mockCompactorFailing) Write(dest string, b BlockReader, mint, maxt int64) (ulid.ULID, error) {
-	if len(c.blocks) > 0 {
-		return ulid.ULID{}, fmt.Errorf("the compactor already did one block so forcefuly failing")
+	if len(c.blocks) >= c.max {
+		return ulid.ULID{}, fmt.Errorf("the compactor already did the maximum allowed blocks so it is time to fail")
 	}
+
 	entropy := rand.New(rand.NewSource(time.Now().UnixNano()))
 	uid := ulid.MustNew(ulid.Now(), entropy)
 	meta := &BlockMeta{
 		Version: 2,
 		ULID:    uid,
 	}
 
-	fmt.Println(dest)
 	block := createEmptyBlock(c.t, filepath.Join(dest, meta.ULID.String()), meta)
 	c.blocks = append(c.blocks, block)
+
+	// Now check that all expected blocks are actually persisted on disk.
+	// This way we make sure that the we have some blocks that are supposed to be removed.
+	var expectedBlocks []string
```

Comment:
My idea was to have something like `c.originalBlocks`. And inside `if len(c.blocks) >= c.max`, we can check that `c.originalBlocks` and `actualBlockDirs` don't match. But seems like this also works!

Reply:
hm, still don't get your idea. , feel free to add another PR with your idea if it is more logical don't mind at all.